pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,119 @@
+package writelimiter
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+// ErrResourcesExhausted is returned when no more BlobWriters can be created
+// because of the limit.
+var ErrResourcesExhausted = errors.New("writers limit is reached")
+
+type writerBouncer struct {
+	sem  *semaphore
+	once sync.Once
+}
+
+func newWriterBouncer(sem *semaphore) (*writerBouncer, error) {
+	if !sem.TryDown() {
+		return nil, ErrResourcesExhausted
```

Comment:
This has to cause a retry by clients, so this isn't the right error.  Need to sync with ben to find an error that he already retries on.  It's probably going to have to be a structured error that is returned all the way to the client.  429 is the "rate limit exceeded, try again" error, so that's what this should be returning (or a higher level needs to convert an error like this into a 429).